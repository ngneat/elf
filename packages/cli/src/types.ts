import { SourceFile } from 'ts-morph';

export interface Options {
  storeName: string;
  features: Array<Features>;
  crud: Array<
    'addEntities' | 'updateEntities' | 'deleteEntities' | 'setEntities'
  >;
  path: string;
  idKey: string;
  template?: 'class' | 'functions';
  inlineStoreInClass?: boolean;
  hooks?: Array<Hooks>;
}

export interface Hooks {
  post?(options: {
    sourceFile: SourceFile;
    options: Omit<Options, 'hooks'>;
    repoName: string;
  }): void;
}

export interface GlobalConfig {
  cli?: {
    repoTemplate?: Options['template'];
    inlineStoreInClass?: Options['inlineStoreInClass'];
    idKey?: Options['idKey'];
    repoLibrary?: string;
    plugins?: string[];
    fuzzypath?: {
      rootPath?: string;
      excludeFilter?: Function;
      excludePath?: Function;
    };
  };
}

export type Features = typeof baseFeatures[number]['value'];

export const baseClassStorePlaces = [
  { name: 'Outside of a class', value: false },
  { name: 'Inside a class constructor', value: true },
] as const;

export const baseFeatures = [
  { name: 'Props', value: 'withProps' },
  { name: 'Entities', value: 'withEntities' },
  {
    name: 'UIEntities',
    value: 'withUIEntities',
  },
  {
    name: 'Active Id',
    value: 'withActiveId',
  },
  {
    name: 'Active Ids',
    value: 'withActiveIds',
  },
  {
    name: 'Requests Cache',
    value: 'withRequestsCache',
  },
  {
    name: 'Requests Status',
    value: 'withRequestsStatus',
  },
] as const;

export const DEFAULT_ID_KEY = 'id';
