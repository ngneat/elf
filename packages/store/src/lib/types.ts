export type OrArray<T> = T[] | T;
export type StateOf<S extends (...args: any) => any> = ReturnType<S>['props'];
export type Query<S, R> = (state: S) => R;
export type NotVoid<R> = R extends void ? never : R;
