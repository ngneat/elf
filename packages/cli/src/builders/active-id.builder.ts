import { FeatureBuilder } from './feature-builder';
import { Features } from '../types';
import { StructureKind, ClassMemberStructures } from 'ts-morph';
import { factory } from 'typescript';

export class ActiveIdBuilder extends FeatureBuilder {
  static supports(featureName: Features): boolean {
    return featureName === 'withActiveId';
  }

  getPropsFactory() {
    return factory.createCallExpression(
      factory.createIdentifier('withActiveId'),
      undefined,
      []
    );
  }

  run() {
    this.addImport(
      ['withActiveId', 'selectActiveEntity', 'setActiveId'],
      '@ngneat/elf-entities'
    );

    const initializer = `${this.storeVariableName}.pipe(selectActiveEntity())`;
    const memberData: ClassMemberStructures = {
      name: `active${this.storeSingularNames.className}$`,
      kind: StructureKind.Property,
    };

    if (this.isStoreInlinedInClass) {
      this.repo.insertProperty(0, {
        ...memberData,
        type: `Observable<${this.storeSingularNames.className} | undefined>`,
      });

      this.repoConstructor?.addStatements(
        `this.${memberData.name} = ${initializer};`
      );
    } else {
      this.repo.insertMember(0, {
        ...memberData,
        initializer,
      });
    }

    this.repo.addMember({
      kind: StructureKind.Method,
      name: this.isFunctionsTpl
        ? `setActive${this.storeNames.className}Id`
        : `setActiveId`,
      parameters: [
        {
          name: 'id',
          type: `${this.storeSingularNames.className}['${this.idKey}']`,
        },
      ],
      statements: [`${this.storeVariableName}.update(setActiveId(id));`],
    });
  }
}
