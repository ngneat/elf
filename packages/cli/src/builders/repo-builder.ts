import {
  ClassDeclaration,
  ConstructorDeclaration,
  printNode,
  Project,
  QuoteKind,
  Scope,
  SourceFile,
  StructureKind,
  SyntaxKind,
  VariableDeclarationKind,
} from 'ts-morph';
import { CallExpression, factory, ScriptTarget } from 'typescript';
import { Options } from '../types';
import { names, resolveStoreVariableName } from '../utils';
import { ActiveIdBuilder } from './active-id.builder';
import { ActiveIdsBuilder } from './active-ids.builder';
import { EntitiesBuilder } from './entities.builder';
import { PropsBuilder } from './props.builder';
import { RequestsCacheBuilder } from './requests-cache.builder';
import { RequestsStatusBuilder } from './requests-status.builder';
import { UIEntitiesBuilder } from './ui-entities.builder';

export function createRepo(options: Options) {
  const { storeName } = options;
  const storeNames = names(storeName);
  const isFunctionTpl = !options.template || options.template === 'functions';
  const isStoreInlinedInClass = !isFunctionTpl && options.inlineStoreInClass;

  const project = new Project({
    manipulationSettings: {
      quoteKind: QuoteKind.Single,
    },
    compilerOptions: {
      target: ScriptTarget.ES2015,
    },
  });

  const sourceFile = project.createSourceFile(`repo.ts`, ``);

  const repoName = `${storeNames.className}Repository`;
  const repoClassDec = sourceFile.addClass({
    name: repoName,
    isExported: true,
  });
  let repoClassDecConstructor: ConstructorDeclaration | undefined;

  if (isStoreInlinedInClass) {
    repoClassDec.addConstructor();
    repoClassDecConstructor = repoClassDec.getConstructors()[0];
  }

  sourceFile.addImportDeclaration({
    moduleSpecifier: '@ngneat/elf',
    namedImports: ['createStore'].map((name) => ({
      kind: StructureKind.ImportSpecifier,
      name,
    })),
  });

  const builders = [
    RequestsCacheBuilder,
    RequestsStatusBuilder,
    ActiveIdBuilder,
    ActiveIdsBuilder,
    EntitiesBuilder,
    PropsBuilder,
    UIEntitiesBuilder,
  ];

  const propsFactories: CallExpression[] = [];

  for (const feature of options.features) {
    for (const Builder of builders) {
      if (Builder.supports(feature)) {
        const instance = new Builder(sourceFile, repoClassDec, options);
        instance.run();
        propsFactories.push(instance.getPropsFactory());
      }
    }
  }

  const storeOpts = factory.createIdentifier(
    `{ name: '${storeNames.propertyName}' }`
  );

  const store = factory.createCallExpression(
    factory.createIdentifier('createStore'),
    undefined,
    [storeOpts, ...propsFactories]
  );

  if (isStoreInlinedInClass && repoClassDecConstructor) {
    addInlineStoreToRepoClass({
      repoClassDec,
      repoClassDecConstructor,
      options,
      store,
      storeNames,
    });
  } else {
    addStoreToRepo({
      repoClassDec,
      options,
      store,
      storeNames,
      isFunctionTpl,
      sourceFile,
    });

    if (isFunctionTpl) {
      toFunctions(sourceFile, repoClassDec);
    }
  }

  if (options.hooks) {
    options.hooks.forEach((h) => h.post?.({ sourceFile, repoName, options }));
  }

  sourceFile.formatText({ indentSize: 2 });

  return sourceFile.getText();
}

function addStoreToRepo({
  repoClassDec: classDec,
  sourceFile,
  options,
  storeNames,
  store,
  isFunctionTpl,
}: {
  repoClassDec: ClassDeclaration;
  sourceFile: SourceFile;
  options: Options;
  storeNames: ReturnType<typeof names>;
  store: CallExpression;
  isFunctionTpl: boolean;
}) {
  const repoPosition = classDec.getChildIndex();

  sourceFile.insertVariableStatement(repoPosition, {
    declarationKind: VariableDeclarationKind.Const,
    isExported: isFunctionTpl,
    declarations: [
      {
        name: resolveStoreVariableName(options.template, storeNames),
        initializer: printNode(store),
      },
    ],
  });
}

function addInlineStoreToRepoClass({
  repoClassDec: classDec,
  repoClassDecConstructor: constructorDec,
  options,
  storeNames,
  store,
}: {
  repoClassDec: ClassDeclaration;
  repoClassDecConstructor: ConstructorDeclaration;
  options: Options;
  storeNames: ReturnType<typeof names>;
  store: CallExpression;
}) {
  const storeName = resolveStoreVariableName(
    options.template,
    storeNames,
    true
  );
  const { propertyIndex, methodIndex } = getPositionsOfInlineStoreDeclarations(
    classDec,
    constructorDec
  );

  const createStoreMethodName = 'createStore';

  classDec.insertMethod(methodIndex, {
    name: createStoreMethodName,
    returnType: `typeof store`,
    scope: Scope.Private,
    statements: (writer) => {
      writer.writeLine(`const store = ${printNode(store)};`);
      writer.blankLine();
      writer.writeLine(`return store;`);
    },
  });

  constructorDec.insertStatements(
    0,
    `${storeName} = this.${createStoreMethodName}();`
  );

  const storeProperty = classDec.insertProperty(propertyIndex, {
    name: `${resolveStoreVariableName(options.template, storeNames)}`,
    scope: Scope.Private,
  });

  if (propertyIndex > 0) {
    storeProperty?.prependWhitespace('\n');
  }
}

function toFunctions(sourceFile: SourceFile, classDec: ClassDeclaration) {
  const exported: string[] = [];

  classDec?.getProperties().forEach((p) => {
    exported.push(`export const ${p.getText()}`);
  });

  classDec?.getMethods().forEach((m) => {
    exported.push(`export function ${m.getText()}`);
  });

  classDec?.remove();

  sourceFile.replaceWithText(
    `${sourceFile.getText()}\n ${exported.join('\n\n')}`
  );
}

function getPositionsOfInlineStoreDeclarations(
  classDec: ClassDeclaration,
  constructorDec: ConstructorDeclaration
) {
  const lastPropertyIndex = classDec
    .getLastChildByKind(SyntaxKind.PropertyDeclaration)
    ?.getChildIndex();
  const lastMethodIndex = classDec
    .getLastChildByKind(SyntaxKind.MethodDeclaration)
    ?.getChildIndex();

  return {
    methodIndex: (lastMethodIndex ?? constructorDec.getChildIndex()) + 1,
    propertyIndex: lastPropertyIndex
      ? lastPropertyIndex + 1
      : constructorDec.getChildIndex(),
  };
}
