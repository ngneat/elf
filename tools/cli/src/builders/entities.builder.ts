import { FeatureBuilder } from './feature-builder';
import { DEFAULT_ID_KEY, Features } from '../types';
import { StructureKind } from 'ts-morph';
import { factory } from 'typescript';

export class EntitiesBuilder extends FeatureBuilder {
  static supports(featureName: Features): boolean {
    return featureName === 'withEntities';
  }

  getPropsFactory() {
    const type: any[] = [
      factory.createTypeReferenceNode(
        factory.createIdentifier(this.storeSingularNames.className),
        undefined
      ),
    ];

    const notDefaultId = this.idKey !== DEFAULT_ID_KEY;

    if (notDefaultId) {
      type.push(
        factory.createLiteralTypeNode(
          factory.createStringLiteral(this.idKey, true)
        )
      );
    }

    let props: any[] = [];

    if (notDefaultId) {
      props = [
        factory.createObjectLiteralExpression(
          [
            factory.createPropertyAssignment(
              factory.createIdentifier('idKey'),
              factory.createStringLiteral(this.idKey, true)
            ),
          ],
          false
        ),
      ];
    }

    return factory.createCallExpression(
      factory.createIdentifier('withEntities'),
      type,
      props
    );
  }

  run() {
    this.addImport(
      ['withEntities', 'selectAll', ...this.options.crud],
      '@ngneat/elf-entities'
    );

    this.sourceFile.insertInterface(this.getLastImportIndex() + 1, {
      name: this.storeSingularNames.className,
      isExported: true,
      properties: [
        {
          name: this.idKey,
          type: 'number',
        },
      ],
    });

    this.repo.addMember({
      name: `${this.storeNames.propertyName}$`,
      kind: StructureKind.Property,
      initializer: `${this.storeVariableName}.pipe(selectAll())`,
    });

    this.options.crud.forEach((op) => this[op]?.());
  }

  setEntities() {
    this.repo.addMember({
      kind: StructureKind.Method,
      name: `set${this.storeNames.className}`,
      parameters: [
        {
          name: this.storeNames.propertyName,
          type: `${this.storeSingularNames.className}[]`,
        },
      ],
      statements: [
        `${this.storeVariableName}.reduce(setEntities(${this.storeNames.propertyName}));`,
      ],
    });
  }

  addEntities() {
    this.repo.addMember({
      kind: StructureKind.Method,
      name: `add${this.storeSingularNames.className}`,
      parameters: [
        {
          name: this.storeSingularNames.propertyName,
          type: this.storeSingularNames.className,
        },
      ],
      statements: [
        `${this.storeVariableName}.reduce(addEntities(${this.storeSingularNames.propertyName}));`,
      ],
    });
  }

  updateEntities() {
    this.repo.addMember({
      kind: StructureKind.Method,
      name: `update${this.storeSingularNames.className}`,
      parameters: [
        {
          name: 'id',
          type: `${this.storeSingularNames.className}['${this.idKey}']`,
        },
        {
          name: this.storeSingularNames.propertyName,
          type: `Partial<${this.storeSingularNames.className}>`,
        },
      ],
      statements: [
        `${this.storeVariableName}.reduce(updateEntities(id, ${this.storeSingularNames.propertyName}));`,
      ],
    });
  }

  deleteEntities() {
    this.repo.addMember({
      kind: StructureKind.Method,
      name: `delete${this.storeSingularNames.className}`,
      parameters: [
        {
          name: 'id',
          type: `${this.storeSingularNames.className}['${this.idKey}']`,
        },
      ],
      statements: [`${this.storeVariableName}.reduce(deleteEntities(id));`],
    });
  }
}
