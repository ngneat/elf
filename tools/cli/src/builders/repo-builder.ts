import {
  ClassDeclaration,
  printNode,
  Project,
  QuoteKind,
  SourceFile,
  StructureKind,
  VariableDeclarationKind,
} from 'ts-morph';
import { CallExpression, factory, ScriptTarget } from 'typescript';
import { RequestsCacheBuilder } from './requests-cache.builder';
import { ActiveIdsBuilder } from './active-ids.builder';
import { EntitiesBuilder } from './entities.builder';
import { UIEntitiesBuilder } from './ui-entities.builder';
import { RequestsStatusBuilder } from './requests-status.builder';
import { ActiveIdBuilder } from './active-id.builder';
import { PropsBuilder } from './props.builder';
import { Options } from '../types';
import { names, resolveStoreVariableName } from '../utils';

export function createRepo(options: Options) {
  const { storeName } = options;
  const storeNames = names(storeName);
  const isFunctionTpl = !options.template || options.template === 'functions';

  const project = new Project({
    manipulationSettings: {
      quoteKind: QuoteKind.Single,
    },
    compilerOptions: {
      target: ScriptTarget.ES2015,
    },
  });

  const sourceFile = project.createSourceFile(`repo.ts`, ``);

  const repoName = `${storeNames.className}Repository`;
  const repoClassDec = sourceFile.addClass({
    name: repoName,
    isExported: true,
  });

  sourceFile.addImportDeclaration({
    moduleSpecifier: '@ngneat/elf',
    namedImports: ['Store', 'createState'].map((name) => ({
      kind: StructureKind.ImportSpecifier,
      name,
    })),
  });

  const builders = [
    RequestsCacheBuilder,
    RequestsStatusBuilder,
    ActiveIdBuilder,
    ActiveIdsBuilder,
    EntitiesBuilder,
    PropsBuilder,
    UIEntitiesBuilder,
  ];

  const propsFactories: CallExpression[] = [];

  for (const feature of options.features) {
    for (const Builder of builders) {
      if (Builder.supports(feature)) {
        const instance = new Builder(sourceFile, repoClassDec, options);
        instance.run();
        propsFactories.push(instance.getPropsFactory());
      }
    }
  }

  const state = factory.createCallExpression(
    factory.createIdentifier('createState'),
    undefined,
    propsFactories
  );

  const repoPosition = repoClassDec.getChildIndex();

  sourceFile.insertVariableStatement(repoPosition, {
    declarationKind: VariableDeclarationKind.Const,
    isExported: isFunctionTpl,
    declarations: [
      {
        name: resolveStoreVariableName(options.template, storeNames),
        initializer: `new Store({ name: '${storeNames.propertyName}', state, config })`,
      },
    ],
  });

  sourceFile.insertVariableStatement(repoPosition, {
    declarationKind: VariableDeclarationKind.Const,
    declarations: [
      {
        name: '{ state, config }',
        initializer: printNode(state),
      },
    ],
  });

  if (isFunctionTpl) {
    toFunctions(sourceFile, repoClassDec);
  }

  if (options.hooks) {
    options.hooks.forEach((h) => h.post?.({ sourceFile, repoName, options }));
  }

  sourceFile.formatText({ indentSize: 2 });

  return sourceFile.getText();
}

function toFunctions(sourceFile: SourceFile, classDec: ClassDeclaration) {
  const exported: string[] = [];

  classDec?.getProperties().forEach((p) => {
    exported.push(`export const ${p.getText()}`);
  });

  classDec?.getMethods().forEach((m) => {
    exported.push(`export function ${m.getText()}`);
  });

  classDec?.remove();

  sourceFile.replaceWithText(
    `${sourceFile.getText()}\n ${exported.join('\n\n')}`
  );
}
