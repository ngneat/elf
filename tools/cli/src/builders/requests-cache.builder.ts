import { FeatureBuilder } from './feature-builder';
import { Features } from '../types';
import { factory } from 'typescript';
import { StructureKind } from 'ts-morph';

export class RequestsCacheBuilder extends FeatureBuilder {
  static supports(featureName: Features): boolean {
    return featureName === 'withRequestsCache';
  }

  getPropsFactory() {
    return factory.createCallExpression(
      factory.createIdentifier('withRequestsCache'),
      undefined,
      []
    );
  }

  run() {
    this.addImport(
      ['withRequestsCache', 'updateRequestsCache', 'CacheState'],
      '@ngneat/elf-requests'
    );

    this.repo.addMember({
      name: `updateRequestCache`,
      kind: StructureKind.Method,
      parameters: [{ name: 'state', type: 'CacheState' }],
      statements: `${this.storeVariableName}.reduce(updateRequestsCache({ key: state }));`,
    });
  }
}
