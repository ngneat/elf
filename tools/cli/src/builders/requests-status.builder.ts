import { FeatureBuilder } from './feature-builder';
import { Features } from '../types';
import { StructureKind } from 'ts-morph';
import { factory } from 'typescript';

export class RequestsStatusBuilder extends FeatureBuilder {
  static supports(featureName: Features): boolean {
    return featureName === 'withRequestsStatus';
  }

  getPropsFactory() {
    return factory.createCallExpression(
      factory.createIdentifier('withRequestsStatus'),
      undefined,
      []
    );
  }

  run() {
    this.addImport(
      [
        'withRequestsStatus',
        'selectRequestStatus',
        'updateRequestsStatus',
        'StatusState',
      ],
      '@ngneat/elf-requests'
    );

    this.repo.insertMember(0, {
      name: `status$`,
      kind: StructureKind.Property,
      initializer: `${this.storeVariableName}.pipe(selectRequestStatus('key'))`,
    });

    this.repo.addMember({
      name: `updateRequestStatus`,
      kind: StructureKind.Method,
      parameters: [{ name: 'state', type: 'StatusState' }],
      statements: `${this.storeVariableName}.reduce(updateRequestsStatus({ key: state }));`,
    });
  }
}
