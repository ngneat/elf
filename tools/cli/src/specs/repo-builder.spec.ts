import { createRepo } from '../builders/repo-builder';
import { baseFeatures } from '../types';

describe('Repo Builder', () => {
  it('withProps', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withProps'],
      crud: [],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withProps');
  });

  it('withRequestsCache', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withRequestsCache'],
      crud: [],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withRequestsCache');
  });

  it('withRequestsStatus', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withRequestsStatus'],
      crud: [],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withRequestsStatus');
  });

  it('withActiveId', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withActiveId'],
      crud: [],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withActiveId');
  });

  it('withActiveIds', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withActiveIds'],
      crud: [],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withActiveIds');
  });

  it('withEntities', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withEntities'],
      crud: ['addEntities', 'updateEntities', 'deleteEntities'],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withEntities');
  });

  it('withEntities idKey', () => {
    const output = createRepo({
      idKey: '_id',
      features: ['withEntities'],
      crud: ['addEntities', 'updateEntities', 'deleteEntities'],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withEntities');
  });

  it('withUIEntities', () => {
    const output = createRepo({
      idKey: 'id',
      features: ['withUIEntities'],
      crud: [],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('withUIEntities');
  });

  it('all', () => {
    const output = createRepo({
      idKey: 'id',
      features: baseFeatures.map(({ value }) => value),
      crud: ['deleteEntities', 'addEntities', 'updateEntities'],
      storeName: 'todos',
      path: '',
    });

    expect(output).toMatchSnapshot('all');
  });

  it('should support function template', () => {
    const output = createRepo({
      idKey: 'id',
      features: baseFeatures.map(({ value }) => value),
      crud: ['deleteEntities', 'addEntities', 'updateEntities'],
      storeName: 'todos',
      path: '',
      template: 'functions',
    });

    expect(output).toMatchSnapshot('all');
  });
});
