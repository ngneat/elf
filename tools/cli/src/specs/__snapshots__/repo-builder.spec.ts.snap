// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repo Builder all: all 1`] = `
"import { Store, createState, withProps } from '@ngneat/elf';
import { withEntities, selectAll, deleteEntities, addEntities, updateEntities, withUIEntities, withActiveId, selectActiveEntity, setActiveId, withActiveIds, selectActiveEntities, toggleActiveIds } from '@ngneat/elf-entities';
import { withRequestsCache, updateRequestsCache, CacheState, withRequestsStatus, selectRequestStatus, updateRequestsStatus, StatusState } from '@ngneat/elf-requests';

export interface TodoUI {
  id: number;
}

export interface Todo {
  id: number;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface TodosProps {
}

const { state, config } = createState(withProps<TodosProps>({}), withEntities<Todo>(), withUIEntities<TodoUI>(), withActiveId(), withActiveIds(), withRequestsCache(), withRequestsStatus());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  status$ = store.pipe(selectRequestStatus('key'));
  activeTodos$ = store.pipe(selectActiveEntities());
  activeTodo$ = store.pipe(selectActiveEntity());
  todos$ = store.pipe(selectAll());

  deleteTodo(id: Todo['id']) {
    store.reduce(deleteEntities(id));
  }

  addTodo(todo: Todo) {
    store.reduce(addEntities(todo));
  }

  updateTodo(id: Todo['id'], todo: Partial<Todo>) {
    store.reduce(updateEntities(id, todo));
  }

  setActiveId(id: Todo['id']) {
    store.reduce(setActiveId(id));
  }

  toggleActiveIds(ids: Array<Todo['id']>) {
    store.reduce(toggleActiveIds(ids));
  }

  updateRequestCache(state: CacheState) {
    store.reduce(updateRequestsCache({ key: state }));
  }

  updateRequestStatus(state: StatusState) {
    store.reduce(updateRequestsStatus({ key: state }));
  }
}
"
`;

exports[`Repo Builder should support function template: all 1`] = `
"import { Store, createState, withProps } from '@ngneat/elf';
import { withEntities, selectAll, deleteEntities, addEntities, updateEntities, withUIEntities, withActiveId, selectActiveEntity, setActiveId, withActiveIds, selectActiveEntities, toggleActiveIds } from '@ngneat/elf-entities';
import { withRequestsCache, updateRequestsCache, CacheState, withRequestsStatus, selectRequestStatus, updateRequestsStatus, StatusState } from '@ngneat/elf-requests';

export interface TodoUI {
  id: number;
}

export interface Todo {
  id: number;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface TodosProps {
}

const { state, config } = createState(withProps<TodosProps>({}), withEntities<Todo>(), withUIEntities<TodoUI>(), withActiveId(), withActiveIds(), withRequestsCache(), withRequestsStatus());
export const todosStore = new Store({ name: 'todos', state, config });

export const status$ = todosStore.pipe(selectRequestStatus('key'));

export const activeTodos$ = todosStore.pipe(selectActiveEntities());

export const activeTodo$ = todosStore.pipe(selectActiveEntity());

export const todos$ = todosStore.pipe(selectAll());

export function deleteTodo(id: Todo['id']) {
  todosStore.reduce(deleteEntities(id));
}

export function addTodo(todo: Todo) {
  todosStore.reduce(addEntities(todo));
}

export function updateTodo(id: Todo['id'], todo: Partial<Todo>) {
  todosStore.reduce(updateEntities(id, todo));
}

export function setActiveId(id: Todo['id']) {
  todosStore.reduce(setActiveId(id));
}

export function toggleActiveIds(ids: Array<Todo['id']>) {
  todosStore.reduce(toggleActiveIds(ids));
}

export function updateRequestCache(state: CacheState) {
  todosStore.reduce(updateRequestsCache({ key: state }));
}

export function updateRequestStatus(state: StatusState) {
  todosStore.reduce(updateRequestsStatus({ key: state }));
}
"
`;

exports[`Repo Builder withActiveId: withActiveId 1`] = `
"import { Store, createState } from '@ngneat/elf';
import { withActiveId, selectActiveEntity, setActiveId } from '@ngneat/elf-entities';

const { state, config } = createState(withActiveId());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  activeTodo$ = store.pipe(selectActiveEntity());

  setActiveId(id: Todo['id']) {
    store.reduce(setActiveId(id));
  }
}
"
`;

exports[`Repo Builder withActiveIds: withActiveIds 1`] = `
"import { Store, createState } from '@ngneat/elf';
import { withActiveIds, selectActiveEntities, toggleActiveIds } from '@ngneat/elf-entities';

const { state, config } = createState(withActiveIds());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  activeTodos$ = store.pipe(selectActiveEntities());

  toggleActiveIds(ids: Array<Todo['id']>) {
    store.reduce(toggleActiveIds(ids));
  }
}
"
`;

exports[`Repo Builder withEntities idKey: withEntities 1`] = `
"import { Store, createState } from '@ngneat/elf';
import { withEntities, selectAll, addEntities, updateEntities, deleteEntities } from '@ngneat/elf-entities';

export interface Todo {
  _id: number;
}

const { state, config } = createState(withEntities<Todo, '_id'>({ idKey: '_id' }));
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  todos$ = store.pipe(selectAll());

  addTodo(todo: Todo) {
    store.reduce(addEntities(todo));
  }

  updateTodo(id: Todo['_id'], todo: Partial<Todo>) {
    store.reduce(updateEntities(id, todo));
  }

  deleteTodo(id: Todo['_id']) {
    store.reduce(deleteEntities(id));
  }
}
"
`;

exports[`Repo Builder withEntities: withEntities 1`] = `
"import { Store, createState } from '@ngneat/elf';
import { withEntities, selectAll, addEntities, updateEntities, deleteEntities } from '@ngneat/elf-entities';

export interface Todo {
  id: number;
}

const { state, config } = createState(withEntities<Todo>());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  todos$ = store.pipe(selectAll());

  addTodo(todo: Todo) {
    store.reduce(addEntities(todo));
  }

  updateTodo(id: Todo['id'], todo: Partial<Todo>) {
    store.reduce(updateEntities(id, todo));
  }

  deleteTodo(id: Todo['id']) {
    store.reduce(deleteEntities(id));
  }
}
"
`;

exports[`Repo Builder withProps: withProps 1`] = `
"import { Store, createState, withProps } from '@ngneat/elf';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface TodosProps {
}

const { state, config } = createState(withProps<TodosProps>({}));
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
}
"
`;

exports[`Repo Builder withRequestsCache: withRequestsCache 1`] = `
"import { Store, createState } from '@ngneat/elf';
import { withRequestsCache, updateRequestsCache, CacheState } from '@ngneat/elf-requests';

const { state, config } = createState(withRequestsCache());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  updateRequestCache(state: CacheState) {
    store.reduce(updateRequestsCache({ key: state }));
  }
}
"
`;

exports[`Repo Builder withRequestsStatus: withRequestsStatus 1`] = `
"import { Store, createState } from '@ngneat/elf';
import { withRequestsStatus, selectRequestStatus, updateRequestsStatus, StatusState } from '@ngneat/elf-requests';

const { state, config } = createState(withRequestsStatus());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
  status$ = store.pipe(selectRequestStatus('key'));

  updateRequestStatus(state: StatusState) {
    store.reduce(updateRequestsStatus({ key: state }));
  }
}
"
`;

exports[`Repo Builder withUIEntities: withUIEntities 1`] = `
"import { Store, createState } from '@ngneat/elf';
import { withUIEntities } from '@ngneat/elf-entities';

export interface TodoUI {
  id: number;
}

const { state, config } = createState(withUIEntities<TodoUI>());
const store = new Store({ name: 'todos', state, config });

export class TodosRepository {
}
"
`;
